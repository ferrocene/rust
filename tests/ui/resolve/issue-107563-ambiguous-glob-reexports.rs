//@ revisions: edition2015 edition2018
//@ [edition2015] edition: 2015
//@ [edition2018] edition: 2018..
#![deny(ambiguous_glob_reexports)]

pub mod foo {
    pub type X = u8;
}

pub mod bar {
    pub type X = u8;
    pub type Y = u8;
}

pub use foo::*;
//~^ ERROR ambiguous glob re-exports
pub use bar::*;

mod ambiguous {
    mod m1 { pub type A = u8; }
    mod m2 { pub type A = u8; }
    pub use self::m1::*;
    //[edition2015]~^ ERROR ambiguous glob re-exports
    pub use self::m2::*;
}

pub mod single {
    pub use ambiguous::A;
    //[edition2015]~^ ERROR `A` is ambiguous
    //[edition2018]~^^ ERROR unresolved import `ambiguous`
}

pub mod glob {
    pub use ambiguous::*;
    //[edition2018]~^ ERROR unresolved import `ambiguous`
}

pub fn main() {}
