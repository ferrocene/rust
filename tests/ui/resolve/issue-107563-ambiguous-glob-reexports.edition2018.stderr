error[E0432]: unresolved import `ambiguous`
  --> $DIR/issue-107563-ambiguous-glob-reexports.rs:28:13
   |
LL |     pub use ambiguous::A;
   |             ^^^^^^^^^ help: a similar path exists: `crate::ambiguous`
   |
   = note: `use` statements changed in Rust 2018; read more at <https://doc.rust-lang.org/edition-guide/rust-2018/module-system/path-clarity.html>

error[E0432]: unresolved import `ambiguous`
  --> $DIR/issue-107563-ambiguous-glob-reexports.rs:34:13
   |
LL |     pub use ambiguous::*;
   |             ^^^^^^^^^ help: a similar path exists: `crate::ambiguous`
   |
   = note: `use` statements changed in Rust 2018; read more at <https://doc.rust-lang.org/edition-guide/rust-2018/module-system/path-clarity.html>

error: ambiguous glob re-exports
  --> $DIR/issue-107563-ambiguous-glob-reexports.rs:15:9
   |
LL | pub use foo::*;
   |         ^^^^^^ the name `X` in the type namespace is first re-exported here
LL |
LL | pub use bar::*;
   |         ------ but the name `X` in the type namespace is also re-exported here
   |
note: the lint level is defined here
  --> $DIR/issue-107563-ambiguous-glob-reexports.rs:4:9
   |
LL | #![deny(ambiguous_glob_reexports)]
   |         ^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 3 previous errors

For more information about this error, try `rustc --explain E0432`.
